File: python/easy/1880_Check_if_Word_Equals_Summation_of_Two_Words.py
from typing import *
class Solution:
    # 32 ms, faster than 40.00% of Python3 online submissions for Check if Word Equals Summation of Two Words.
    # 14.1 MB, less than 100.00% of Python3 online submissions for Check if Word Equals Summation of Two Words.
    def isSumEqual(self, firstWord: str, secondWord: str, targetWord: str) -> bool:
        def word2value(word):
            ans = 0
            for letter in word:
                ans = ans*10 + (ord(letter) - ord("a")) 
            return ans
        word_value_dict = {i:word2value(word) for i, word in enumerate([firstWord, secondWord, targetWord])}
        return word_value_dict[0] + word_value_dict[1] == word_value_dict[2]

    #  28 ms, faster than 100.00% of Python3 online submissions for Check if Word Equals Summation of Two Words.
    # 28 ms, faster than 100.00% of Python3 online submissions for Check if Word Equals Summation of Two Words.
    def isSumEqual(self, firstWord: str, secondWord: str, targetWord: str) -> bool:
        mapping = {
            'a':'0', 'b':'1', 'c':'2', 'd':'3', 'e':'4', 'f':'5', 'g':'6', 'h':'7', 'i':'8', 'j':'9'
        }
        def decoding(s):
            nonlocal mapping
            return int(''.join(list(map((lambda i:mapping[i]),list(s)))))
        
        return decoding(firstWord) + decoding(secondWord) == decoding(targetWord)


if __name__ == "__main__":
    so = Solution()
    print(so.isSumEqual(firstWord = "aaa", secondWord = "a", targetWord = "aab"))

File: c/easy/1880_Check_if_Word_Equals_Summation_of_Two_Words.c
#include <stdio.h>
#include <stdlib.h>
#include <iso646.h>

// 0 ms, faster than 100.00% of C online submissions for Check if Word Equals Summation of Two Words.
// 5.5 MB, less than 65.91% of C online submissions for Check if Word Equals Summation of Two Words.
_Bool isSumEqual(char * firstWord, char * secondWord, char * targetWord){
    long nums[3] = {0};
    char * strings[3] = {firstWord, secondWord, targetWord};
    int j = 0;
    int k = 0;
    for (int i = 0; i < 3; i++){
        k = 0;
        while (*(strings[j]+k) != '\0'){
            nums[i] = nums[i] *10 + *(strings[j]+k)  - 'a';
            k++;
        }
        j++;
    }
    return nums[0] + nums[1] == nums[2];
}


int main(void){
    char s1[10], s2[10], s3[10];
    scanf("%s", s1);
    scanf("%s", s2);
    scanf("%s", s3);
    printf("%d\n", isSumEqual(s1,s2, s3));

    return 0;
}